modules/broker-core/src/main/java/io/ballerina/messaging/broker/core/BindingsRegistry.java

```diff
@@ -43,38 +51,58 @@
 
     private final BindingDeleteListener bindingDeleteListener;
 
+    private List<BindingsRegistryListener> bindingsRegistryListeners;
+
+    private final ReadWriteLock lock;
+
     BindingsRegistry(Exchange exchange, BindingDao bindingDao) {
         this.bindingPatternToBindingsMap = new HashMap<>();
         this.exchange = exchange;
         this.bindingDao = bindingDao;
         this.unmodifiableBindingSetView = Collections.unmodifiableMap(bindingPatternToBindingsMap);
         bindingDeleteListener = new BindingDeleteListener();
+        bindingsRegistryListeners = new ArrayList<>();
+        lock = new ReentrantReadWriteLock();
     }
 
     void bind(QueueHandler queueHandler, String bindingKey, FieldTable arguments) throws BrokerException,
-                                                                                         ValidationException {
-        BindingSet bindingSet = bindingPatternToBindingsMap.computeIfAbsent(bindingKey, k -> new BindingSet());
-        Queue queue = queueHandler.getQueue();
-        Binding binding = new Binding(queue, bindingKey, arguments);
-        boolean success = bindingSet.add(binding);
-
-        if (success) {
-            queueHandler.addBinding(binding, bindingDeleteListener);
-            if (queue.isDurable()) {
-                bindingDao.persist(exchange.getName(), binding);
+            ValidationException {
+        lock.writeLock().lock();
```

Comment:
We don't need a lock here since BrokerImpl holds a lock before propagating the call here.

Reply:
Fixed in cd6b801